pennylane/qnodes/base.py

```diff
@@ -474,8 +489,27 @@ def _construct(self, args, kwargs):
         """
         # pylint: disable=attribute-defined-outside-init, too-many-branches, too-many-statements
 
-        if self.arg_vars is None or self.kwarg_vars is None:
-            self.arg_vars, self.kwarg_vars = self._make_variables(args, kwargs)
+        def equal_dicts(a, b):
+            """Return True iff a and b are equal dicts."""
+            if b is None:
+                return False
+
+            if len(a) != len(b):
+                return False
+
+            for key, val in a.items():
+                if np.any(b[key] != val):
+                    return False
+            return True
+
+        if equal_dicts(kwargs, self.last_aux_args):
+            return
```

Comment:
The main change in the logic is here. Immutable circuits should be constructed only once, mutable circuits should only be reconstructed when the auxiliary args (kwargs) change.

Reply:
I could not simply compare the dicts with `==`, because it won't work when the dicts contain `arrays` as values.

Reply:
Currently this logic seems to assume, that the mutability of the QNode merely depends on the auxiliary arguments passed to the qfunc.

Consider the following case:
```python
def circuit(x, n=1): 
    for k in range(len(x)): 
        for j in range(len(x[k])): 
            qml.RX(x[k][j], wires=0) 
            print(len(x[k])) 
            print(x[k][j].val) 
    return qml.expval(qml.PauliZ(wires=0)) 

node = BaseQNode(circuit, dev, mutable=True) 

print(node([[0.1], [0.2]], n=1)) 
print(node([[0.1], [0.2, 0.3]], n=1))
```
The outputs are the following:
```
1
0.1
1
0.2
0.9553364891256061
0.9553364891256061
```
The discrepency comes from the fact that the change in the length of the positional arguments was not considered.

Reply:
An attempt to capture this as well was attempted in the original form of #505

Reply:
@antalszava is right --- we did not account in the design or the test suite for mutability based on differentiable arguments, but it did previously work in limited cases (i.e., by considering the length of a positional arg). This is what the QAOA tutorial was doing, and this is what I've seen other users do. Our options are either:

1. Breaking usability change - going forward, only auxiliary arguments can mutate a QNode

2. Keep the existing behaviour.

I would argue for (2), as it is intuitive --- looking at the above code snippet as a new user, I would expect that to work.

Reply:
As far as I can see, in `tutorial_qaoa_maxcut.py` the circuit structure does not depend on positional args (gammas, betas), only on the auxiliary args (edge, n_layers). `gammas` and `betas` are only used as `Operator` arguments, not in flow control structures.

What instead happens is the auxiliary arg `n_layers` determines the length of `gammas` and `betas`, which is consistent with the qfunc rule `"only auxiliary args may affect the structure of the circuit"`.

I'd strongly argue for keeping this rule, since without it we have no option but to reconstruct the entire circuit *each time* any of the arguments change, which would make differentiation super inefficient (and might outright break the parameter-shift rule, since it assumes differentiability which is no longer guaranteed if the differentiable parameters can be used in flow control.)

The rule is not currently enforced by the code, it's the user's responsibility to make sure the node is called each time with the correct positional args. But maybe there could be simple, lightweight ways of checking that the args are ok.

Reply:
Recently I used the following code, and it worked:
```python
@qml.qnode(dev)
def circuit(weights_list):
    for weights in weights_list:
        template(weights)

    return qml.expval(qml.PauliZ(0))
```
I don't see why I should need to provide an additional kwarg that tells me how many `weights` are in the `weights_list`

Reply:
Do you mean `len(weights_list)`? That may not be enough, since any of the list elements could be another list, or array, that could change in dimension and contents, which would require creating new VariableRefs.

The above code works fine if the node is always called with a `weights_list` having the same nested structure, but it'd be safer to set the range of the for loop using a constant or an auxiliary parameter.

Reply:
My understanding here is that the tutorial broke because it was depending on the nested structure (`shape`) of the positional argument.

An attempt to aid this was to:
* Convert all positional arguments to an `ndarray`
* For the first run, store the shape of this `ndarray`
* On subsequent runs, check the current shape; if different, then
   * Recreate the `VariableRef`s according to the current shape
   * Store the new shape

This did not work with cases an `ndarray` with shapes like `(2,2,1)` were passed. Therefore, a `try-except` block was created and the `numpy.squeeze` was called first on such arrays.

To me, this check would seem to cover a subset of the cases that we would like to provide caching for (but not the entire set, as @smite already pointed out with providing examples where only using this condition does not work).

Reply:
The problem of differentiability still remains, in the supermutable case it would be up to the user to guarantee that the circuit is differentiable wrt. the primary args.

Reply:
> the user to guarantee that the circuit is differentiable wrt. the primary args.

I'm fine with this as long as it is communicated to the user well